File: code/160/solution.cpp
#include <iostream>
using namespace std;
const long long upper = 1e12;
const long long e5    = 1e5;
long long qpow(long long a, long long b) {
    long long ret = 1;
    while (b) {
        if (b & 1)
            ret = (ret * a) % e5;
        a = (a * a) % e5;
        b >>= 1;
    }
    return ret;
}
long long count(long long a, long long base) { /*{{{*/
    long long ret = 0;
    while (a) {
        a /= base;
        ret += a;
    }
    return ret;
} /*}}}*/
int cnt[e5];
int main() {
    long long ans = 1;
    for (long long a = 1; a <= upper; a <<= 1) {
        for (long long b = 1; a * b <= upper; b *= 5) {
            long long x = upper / (a * b);
            for (long long y = 1; y < e5; y += 2) {
                if (y % 5 == 0)
                    continue;
                cnt[y] += x / e5 + (x % e5 >= y);
            }
        }
    }
    for (long long y = 1; y < e5; y++) {
        ans = (ans * qpow(y, cnt[y])) % e5;
    }
    ans = (ans * qpow(2, count(upper, 2) - count(upper, 5))) % e5;
    cout << ans << endl;
    return 0;
}

File: code/160/solution.py
from random import random
from math import *

upper = int(1e12)
e5 = int(1e5)


def qpow(a, b):
    ret = 1
    while b > 0:
        if (b & 1) > 0:
            ret = ret * a % e5
        a = a * a % e5
        b >>= 1
    return ret


def count(a, base):
    ret = 0
    while a > 0:
        a = int(a // base)
        ret += a
    return ret


cnt = [0] * e5

if __name__ == "__main__":
    ans = 1
    a = 1
    while a <= upper:
        b = 1
        while a * b <= upper:
            x = upper // (a * b)
            for y in range(1, e5, 2):
                if y % 5 == 0:
                    continue
                cnt[y] += x // e5
                if x % e5 >= y:
                    cnt[y] += 1
            b *= 5
        a <<= 1
    for y in range(1, e5):
        ans = ans * qpow(y, cnt[y]) % e5
    ans = ans * qpow(2, count(upper, 2) - count(upper, 5)) % e5
    print(ans)
